plugin/src/main/java/io/github/guegse/foreachstream/plugin/ForeachStreamPlugin.java

```diff
@@ -31,29 +39,56 @@ public boolean autoStart() {
 
     @Override
     public void init(JavacTask task, String... args) {
-        com.sun.tools.javac.util.Context javacContext = ((BasicJavacTask) task).getContext();
+        Context javacContext = ((BasicJavacTask) task).getContext();
         treeMaker = TreeMaker.instance(javacContext);
         names = Names.instance(javacContext);
         trees = Trees.instance(task);
         task.addTaskListener(this);
+        subs = new HashMap<>();
+        this.log = Log.instance(javacContext);
         availableMethodsToTheirClasses = new HashMap<>();
         for (Class<?> clazz : ForeachStreamClasses.CLASSES) {
             for (Method method : clazz.getMethods()) {
                 if (Modifier.isPublic(method.getModifiers()) && Modifier.isStatic(method.getModifiers())) {
-                    availableMethodsToTheirClasses.put(method.getName(), clazz.getName());
+                    availableMethodsToTheirClasses.put(method.getName(), new Pair<>(clazz.getName(), method.getParameterTypes()));
                 }
             }
         }
+        options = Options.instance(javacContext);
+        this.statistics = new Statistics();
+        for(var option : options.keySet()) {
+            if(option.startsWith("-ArootProjectPath=")) {
+                this.rootProjectPath = option.substring(option.indexOf("=") + 1);
+            } else if(option.startsWith("-AprojectPath")) {
+                this.projectPath = option.substring(option.indexOf("=") + 1);
+            } else if(option.startsWith("-AtaskName")) {
+                this.taskName = option.substring(option.indexOf("=") + 1);
+            }
+        }
     }
 
     @Override
     public void finished(TaskEvent e) {
-        if (e.getKind() != TaskEvent.Kind.PARSE) {
-            return;
+        DebugOutput debugOutput = new DebugOutput(e.getCompilationUnit(), trees, true);
+        if (e.getKind() == TaskEvent.Kind.PARSE) {
+            CompilationUnitTree compilationUnit = e.getCompilationUnit();
+            Substitution sub = new Substitution(treeMaker, debugOutput, statistics);
+            subs.put(compilationUnit, sub);
+            e.getCompilationUnit().accept(
+                    new TreeScanner(treeMaker, names, availableMethodsToTheirClasses, sub, debugOutput, statistics), null);
+        } else if(e.getKind() == TaskEvent.Kind.ANALYZE) {
+            Substitution sub = subs.get(e.getCompilationUnit());
+            if(sub != null) {
+                sub.substitute();
+            }
+        } else if(e.getKind() == TaskEvent.Kind.COMPILATION) {
+            if(statistics != null) {
```

Comment:
statistics cannot be null at this time, right? Otherwise line 89 would also be problematic